### PR TITLE
Eliminate double JSON.stringify in RSC payload embedding

### LINKED ISSUE #2522: Reduce RSC payload overhead: double JSON.stringify adds ~38KB (24%) unnecessary bloat
## Summary

The current RSC payload embedding pipeline applies `JSON.stringify` **twice** to Flight data before injecting it into the HTML stream. This double serialization causes every `"` character in the Flight payload to be escaped twice (`\"` → `\\\"` ), adding **~38KB of unnecessary overhead** (~24% of the raw Flight data) on a typical product search page with 36 results.

## Root Cause

The serialization happens in two phases:

### Phase 1: `streamingUtils.js` (line 102)

```js
const jsonChunk = JSON.stringify(createResultObject(htmlChunk, consoleReplayScript, renderState));
```

This wraps the raw Flight data (`htmlChunk`) in a JSON envelope:
```json
{"html":"0:\"$Sreact.suspense\"\n1:I[\"16109\",...","consoleReplayScript":"","hasErrors":false,"isShellReady":true}
```

Every `"` in the raw Flight data becomes `\"` (+1 byte each). With ~18,883 double quotes in a typical Flight payload, this adds **~21.7KB**.

### Phase 2: `injectRSCPayload.js` (line 38)

```js
return createScriptTag(`(${cacheKeyJSArray(cacheKey)}).push(${JSON.stringify(chunk)})`);
```

This re-stringifies the already-JSON string for embedding in a `<script>` tag. Now every `\"` from Phase 1 becomes `\\\"` (+2 more bytes each), and the outer quotes are also escaped.

This adds another **~38KB** of escaping overhead.

### Total overhead measured

| Stage | Bytes | Overhead vs raw |
|---|---|---|
| Raw Flight data | 159,321 | baseline |
| After 1st `JSON.stringify` | +21,710 | +13.6% |
| After 2nd `JSON.stringify` | +38,058 | +23.9% |
| **Total serialization overhead** | **57,711** | **36.2%** |

## Comparison with Next.js

Next.js uses a **single** `JSON.stringify`:

```js
self.__next_f.push([1, "raw flight data here"])
```

The raw Flight data is stringified only once for embedding in JS. React on Rails could save **~38KB per page** (17.5% of current embedded size) by eliminating the second stringify.

## Proposed Solution

Instead of having `streamingUtils.js` wrap Flight data in a JSON envelope that `injectRSCPayload.js` later re-serializes, consider one of these approaches:

1. **Skip the JSON envelope for RSC payloads** — Pass the Flight data directly to `injectRSCPayload.js` without wrapping it in `createResultObject`. The `consoleReplayScript`, `hasErrors`, and `isShellReady` metadata can be transmitted separately or embedded as a small sidecar.

2. **Single stringify at the embedding point** — Have `injectRSCPayload.js` receive the raw Flight string and only stringify it once for the `.push()` call.

3. **Use `escapeScript()` instead of `JSON.stringify` in Phase 2** — Since the chunk is already valid JSON, it only needs HTML-unsafe characters escaped (e.g., `</script>`, `<!--`), not full JSON re-serialization. The existing `escapeScript()` function already handles this.

## Environment

- `react-on-rails-pro` v16.4.0-rc.5
- `react-on-rails` v16.4.0-rc.3
- Measured on a product search page with 36 result cards, each with descriptions, review snippets, features, and specs

## Reproduction

Save the rendered RSC page HTML and measure the escaping overhead:

```python
import json, re

# Load saved RSC page HTML
html = open('/tmp/rsc-page.html').read()

# Extract push() calls
pushes = re.findall(r'REACT_ON_RAILS_RSC_PAYLOADS.*?\)\.push\((.+?)\)</script>', html)

raw_total = 0
for push_content in pushes:
    # Undo the 2nd JSON.stringify to get the JSON envelope
    envelope_json = json.loads(push_content)
    # Parse the JSON envelope to get raw Flight data
    envelope = json.loads(envelope_json)
    raw_flight = envelope.get('html', '')
    raw_total += len(raw_flight)

print(f"Raw Flight data: {raw_total} bytes")
print(f"Embedded size (with double stringify): measure from HTML")
```

### PR DESCRIPTION
**Pull request by justin808**
## Summary

- **[Pro]** Eliminates the double `JSON.stringify` in RSC payload embedding that was adding ~38KB (~24%) unnecessary overhead per page
- RSC Flight data is now embedded directly as JavaScript expressions in `<script>` tags instead of being re-stringified
- Client-side preloaded payloads are now objects (not strings), skipping the `JSON.parse` overhead during hydration

## Problem

The RSC payload embedding pipeline applied `JSON.stringify` **twice** to Flight data:

1. **Phase 1** (`streamingUtils.ts`): Wraps raw Flight data in a JSON result object envelope
2. **Phase 2** (`injectRSCPayload.ts`): Re-stringifies the already-JSON string for embedding in `<script>` tags

Every `"` in the Flight data gets escaped twice (`\"` → `\\\"`), adding ~38KB on a typical page.

## Solution

Since JSON is a strict subset of JavaScript expressions, embed the JSON result objects directly in the `.push()` call — no re-stringification needed. `escapeScript()` in `createScriptTag()` already handles HTML-unsafe sequences (`</script>`, `<!--`) using JS-compatible escapes.

**Server-side**: Add line buffering in `injectRSCPayload` to split stream chunks on newline boundaries before embedding each complete JSON line directly.

**Client-side**: `createRSCStreamFromPreloadedPayloads` processes `RSCPayloadChunk` objects directly — extracts `html`, replays console logs — without going through `transformRSCStreamAndReplayConsoleLogs` (which previously had to `JSON.parse` each line).

## Test plan

- [x] `injectRSCPayload.test.ts` — 9 tests pass (updated mock data to include newlines matching real format)
- [x] `RSCRequestTracker.test.ts` — 12 tests pass (updated assertion for direct JSON embedding)
- [x] `streamBackpressure.e2e.test.tsx` — all tests pass (backpressure behavior unchanged)
- [x] All Pro RSC tests pass (`test:non-rsc` + `test:rsc`)
- [x] All OSS tests pass (120 tests)
- [x] TypeScript type check passes
- [x] ESLint + Prettier clean

Fixes #2522

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Changes the server→client RSC payload contract and streaming parsing logic (NDJSON line buffering, direct JS embedding), which could affect hydration or error handling if any edge cases are missed, though covered by expanded tests.
> 
> **Overview**
> **Improves Pro RSC hydration performance by removing a redundant `JSON.stringify`** when embedding Flight NDJSON into streamed HTML, by pushing parsed JSON objects directly into `REACT_ON_RAILS_RSC_PAYLOADS` script tags.
> 
> Server-side `injectRSCPayload` now buffers incoming stream chunks into complete NDJSON lines (handling split boundaries / multibyte UTF-8) and validates each line before embedding; client-side preloaded payload handling switches from string payloads + `JSON.parse` to `RSCPayloadChunk` objects, with console replay extracted into a shared `replayConsoleLog` helper.
> 
> Tests were updated and expanded to assert direct `.push({…})` embedding, cover boundary conditions, and ensure malformed NDJSON emits an error rather than injecting invalid JavaScript; CHANGELOG includes the perf note.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit d187d68ef9cf33dacc1361134401f29db5e7ed22. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

## Summary by CodeRabbit

* **Performance**
  * Eliminated redundant JSON serialization for RSC Flight data, reducing payloads (~38KB, ~24% typical) for faster page loads.
* **Reliability**
  * Improved streaming/NDJSON processing, safer handling of preloaded payloads, finalized partial-line flushing, and safer console-replay with nonce support to prevent malformed injections.
* **Tests**
  * Updated and added tests covering raw-object payload embedding, chunk boundaries, multibyte characters, and malformed-NDJSON error handling.

### CODE REVIEW

File: packages/react-on-rails-pro/src/getReactServerComponent.client.ts [hunk 1]
```diff
@@ -91,20 +91,45 @@ const fetchRSC = ({
...
+    .replace(/<\/script>$/i, '');
+  if (replayConsoleCode?.trim() !== '') {
+    const scriptElement = document.createElement('script');
+    if (sanitizedNonceValue) {
+      scriptElement.nonce = sanitizedNonceValue;
+    }
+    scriptElement.textContent = replayConsoleCode;
+    document.body.appendChild(scriptElement);
+  }
+}
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/getReactServerComponent.client.ts]: Addressed in e204b4e — extracted `replayConsoleLog` into `utils.ts` as a shared utility, imported by both `getReactServerComponent.client.ts` and `transformRSCStreamAndReplayConsoleLogs.ts`. The two code paths now share a single implementation.

File: packages/react-on-rails-pro/tests/injectRSCPayload.test.ts [hunk 2]
```diff
@@ -48,36 +48,36 @@ const setupTest = (mockRSC: Readable) => {
...
     );
   });
 
   it('should handle multiple RSC payloads', async () => {
-    const mockRSC = createMockStream(['{"test": "data"}', '{"test": "data2"}']);
+    const mockRSC = createMockStream(['{"test": "data"}\n', '{"test": "data2"}\n']);
     const mockHTML = createMockStream(['<html><body><div>Hello, world!</div></body></html>']);
     const { rscRequestTracker, domNodeId } = setupTest(mockRSC);
 
     const result = injectRSCPayload(mockHTML, rscRequestTracker, domNodeId);
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/tests/injectRSCPayload.test.ts:68]: Addressed in e204b4e — added a test that sends a chunk split across two JSON object boundaries (`'{"test": "data"}\n{"test": "dat'` followed by `'a2"}\n'`) and verifies both objects appear in the output.

File: packages/react-on-rails-pro/src/getReactServerComponent.client.ts [hunk 3]
```diff
@@ -91,20 +91,45 @@ const fetchRSC = ({
...
+      const handleChunk = (chunk: RSCPayloadChunk) => {
+        controller.enqueue(encoder.encode(chunk.html ?? ''));
+        replayConsoleLog(chunk.consoleReplayScript, sanitizedNonceValue);
       };
 
       payloads.forEach(handleChunk);
       // eslint-disable-next-line no-param-reassign
-      payloads.push = (...chunks) => {
+      payloads.push = (...chunks: RSCPayloadChunk[]) => {
         chunks.forEach(handleChunk);
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/getReactServerComponent.client.ts:131]: Declining — this is a false positive. The claimed race condition cannot occur because `DOMContentLoaded` fires only after the HTML parser finishes, which means all inline `<script>` tags (including RSC `.push()` calls injected by `injectRSCPayload`) have already executed. The `injectRSCPayload` transform stream holds the HTML response open until all RSC streams complete, so the closing HTML tags aren't sent (and DOMContentLoaded doesn't fire) until every `.push()` script has been emitted and processed by the browser.

File: packages/react-on-rails-pro/src/getReactServerComponent.client.ts [hunk 4]
```diff
@@ -91,20 +91,45 @@ const fetchRSC = ({
   }
 };
 
-const createRSCStreamFromArray = (payloads: string[]) => {
-  let streamController: ReadableStreamController<string> | undefined;
-  const stream = new ReadableStream<string>({
+function replayConsoleLog(consoleReplayScript: string | undefined, sanitizedNonceValue?: string) {
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/getReactServerComponent.client.ts]: Duplicate of the thread on lines 94–107 — addressed in e204b4e by extracting `replayConsoleLog` into `utils.ts`.

File: packages/react-on-rails-pro/src/getReactServerComponent.client.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/getReactServerComponent.client.ts]: Duplicate of the thread on lines 94–107 — addressed in e204b4e by extracting `replayConsoleLog` into `utils.ts`.

[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/getReactServerComponent.client.ts]: Duplicate of the thread on lines 94–107 — addressed in e204b4e by extracting `replayConsoleLog` into `utils.ts`.

File: packages/react-on-rails-pro/src/getReactServerComponent.client.ts [hunk 5]
```diff
@@ -91,20 +91,45 @@ const fetchRSC = ({
...
+  const replayConsoleCode = (consoleReplayScript ?? '')
+    .trim()
+    .replace(/^<script[^>]*>/i, '')
+    .replace(/<\/script>$/i, '');
+  if (replayConsoleCode?.trim() !== '') {
+    const scriptElement = document.createElement('script');
+    if (sanitizedNonceValue) {
+      scriptElement.nonce = sanitizedNonceValue;
+    }
+    scriptElement.textContent = replayConsoleCode;
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/getReactServerComponent.client.ts]: Declining — this is a minor robustness suggestion. In practice, `document.body.appendChild` will not throw for a valid script element. The console replay is best-effort logging; an uncaught exception here would not affect application behavior. Adding try/catch for a scenario that cannot realistically occur would add noise.

File: packages/react-on-rails-pro/src/injectRSCPayload.ts [hunk 6]
```diff
@@ -251,12 +255,33 @@ export default function injectRSCPayload(
...
+                lastIncompleteLine = '';
+              }
+
+              for (const line of lines) {
+                if (line.trim() !== '') {
+                  const payloadScript = createRSCPayloadChunk(line, rscPayloadKey, sanitizedNonce);
+                  rscPayloadBuffers.push(Buffer.from(payloadScript));
+                }
+              }
+              scheduleFlush();
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/injectRSCPayload.ts]: Declining — this is a documentation nit. The code is straightforward and the surrounding comment on line 258–260 already explains the buffering approach. Adding a comment specifically about scheduleFlush being called per-chunk rather than per-line doesn't add enough value to warrant the diff.

[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/injectRSCPayload.ts]: Declining — this is a minor efficiency nit. An empty scheduleFlush is a no-op (it schedules a setTimeout that flushes an empty buffer). The cost is negligible compared to the stream processing, and adding a guard would reduce readability for zero measurable benefit.

File: CHANGELOG.md [hunk 7]
```diff
@@ -24,6 +24,17 @@ After a release, run `/update-changelog` in Claude Code to analyze commits, writ
...
 ### [Unreleased]
 
+#### Performance
+
+- **[Pro]** **Reduce RSC payload overhead by eliminating double JSON.stringify**: RSC Flight data embedded in
+  HTML stream script tags was being JSON.stringified twice — once when wrapping in the result object envelope,
+  and again when embedding in the `<script>` tag's `.push()` call. The second stringify is now eliminated by
+  embedding JSON directly as a JavaScript expression (JSON is a strict subset of JS). This saves ~38KB (~24%
+  of raw Flight data) on a typical product search page with 36 results.
+  [PR 2835](https://github.com/shakacode/react_on_rails/pull/2835) by
```
[the remark this answers is not in the document]
Comment (justin808) [reply CHANGELOG.md:34]: Already fixed — the CHANGELOG was updated with the actual PR number in commit 51bb3cb84 ("Update CHANGELOG with PR number"). Line 34 now reads `[PR 2835](https://github.com/shakacode/react_on_rails/pull/2835)`.

File: packages/react-on-rails-pro/src/getReactServerComponent.client.ts [hunk 8]
```diff
@@ -91,20 +98,30 @@ const fetchRSC = ({
...
+        controller.enqueue(encoder.encode(chunk.html ?? ''));
+        replayConsoleLog(chunk.consoleReplayScript, sanitizedNonceValue);
       };
 
       payloads.forEach(handleChunk);
       // eslint-disable-next-line no-param-reassign
-      payloads.push = (...chunks) => {
+      payloads.push = (...chunks: RSCPayloadChunk[]) => {
         chunks.forEach(handleChunk);
         return chunks.length;
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/getReactServerComponent.client.ts:132]: Acknowledged — this is technically a contract violation, but since no caller uses the return value and the override doesn't actually store items (it just processes chunks via `handleChunk`), there's no behavioral impact. Leaving as-is to keep the diff focused on the double-stringify fix.

File: packages/react-on-rails-pro/src/injectRSCPayload.ts [hunk 9]
```diff
@@ -251,12 +255,33 @@ export default function injectRSCPayload(
...
+            }
+            if (lastIncompleteLine.trim() !== '') {
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/injectRSCPayload.ts]: The `lastIncompleteLine` flush path is already exercised indirectly by the cross-boundary chunk test added in e204b4e. A dedicated test for this specific branch would be redundant given that coverage. Skipping to keep the PR focused.

File: packages/react-on-rails-pro/src/injectRSCPayload.ts [hunk 10]
```diff
@@ -46,8 +46,12 @@ function createRSCPayloadInitializationScript(cacheKey: string, sanitizedNonce?:
...
 }
 
-function createRSCPayloadChunk(chunk: string, cacheKey: string, sanitizedNonce?: string) {
-  return createScriptTag(`(${cacheKeyJSArray(cacheKey)}).push(${JSON.stringify(chunk)})`, sanitizedNonce);
+function createRSCPayloadChunk(jsonLine: string, cacheKey: string, sanitizedNonce?: string) {
+  // Embed the JSON line directly as a JavaScript expression instead of re-stringifying it.
+  // JSON is a strict subset of JavaScript expressions, so this is always valid JS.
+  // createScriptTag's escapeScript() handles HTML-unsafe sequences (</script>, <!--)
+  // using JS-compatible escape sequences that preserve the parsed value.
+  return createScriptTag(`(${cacheKeyJSArray(cacheKey)}).push(${jsonLine})`, sanitizedNonce);
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/injectRSCPayload.ts:59]: Verified locally: `JSON.stringify` does **not** escape U+2028/U+2029. The character appears as literal bytes (`e2 80 a8`) in the output, not as a `\u2028` escape sequence. This is per the ECMAScript spec — `SerializeJSONString` only escapes code units below U+0020, plus `"` and `\`.

Since the old code (`JSON.stringify(chunk)`) also left U+2028/U+2029 unescaped, the behavior is unchanged. Both old and new code paths place these characters inside JS string literals within the `<script>` tag, which is valid in ES2019+ environments. No action needed.

File: packages/react-on-rails-pro/src/injectRSCPayload.ts [hunk 11]
```diff
@@ -251,12 +255,33 @@ export default function injectRSCPayload(
...
+                  rscPayloadBuffers.push(Buffer.from(payloadScript));
+                }
+              }
+              scheduleFlush();
+            }
+            if (lastIncompleteLine.trim() !== '') {
+              const payloadScript = createRSCPayloadChunk(lastIncompleteLine, rscPayloadKey, sanitizedNonce);
               rscPayloadBuffers.push(Buffer.from(payloadScript));
               scheduleFlush();
             }
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/injectRSCPayload.ts:295]: Addressed in f592091d. I moved TextDecoder scope into each per-stream async task, switched to streaming decode (`decoder.decode(chunk, { stream: true })`), and flush pending bytes with a final `decoder.decode()` before processing the last line. Added a regression test in `injectRSCPayload.test.ts` that splits a multibyte UTF-8 emoji across chunk boundaries to verify chunk-split safety.

File: packages/react-on-rails-pro/src/utils.ts [hunk 12]
```diff
@@ -49,3 +49,18 @@ export const sanitizeNonce = (nonce?: string) => {
   const nonceWithAllowedCharsOnly = nonce?.replace(/[^a-zA-Z0-9+/=_-]/g, '');
   return nonceWithAllowedCharsOnly?.match(/^[a-zA-Z0-9+/_-]+={0,2}$/)?.[0];
 };
+
+export function replayConsoleLog(consoleReplayScript: string | undefined, sanitizedNonce?: string) {
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/utils.ts:53]: Addressed in fb8f4905. Added an explicit browser guard in replayConsoleLog to no-op safely when document is unavailable.

File: packages/react-on-rails-pro/src/injectRSCPayload.ts [hunk 13]
```diff
@@ -251,12 +254,36 @@ export default function injectRSCPayload(
...
+                (typeof chunk === 'string' ? chunk : decoder.decode(chunk, { stream: true }));
+              const lines = decodedChunk.split('\n');
+
+              if (!decodedChunk.endsWith('\n')) {
+                lastIncompleteLine = lines.pop() ?? '';
+              } else {
+                lastIncompleteLine = '';
+              }
+
+              for (const line of lines) {
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/injectRSCPayload.ts:278]: Addressed in fb8f4905 by documenting the NDJSON contract inline and normalizing each line with trim() before emitting payload scripts (improves CRLF/whitespace robustness while keeping the streaming path lightweight).

File: packages/react-on-rails-pro/src/getReactServerComponent.client.ts [hunk 14]
```diff
@@ -91,20 +98,30 @@ const fetchRSC = ({
...
+ * directly as JavaScript expressions, avoiding the double JSON.stringify overhead.
+ * This function extracts the html field and replays console logs from each chunk.
+ */
+const createRSCStreamFromPreloadedPayloads = (payloads: RSCPayloadChunk[], cspNonce?: string) => {
+  const encoder = new TextEncoder();
+  const sanitizedNonceValue = sanitizeNonce(cspNonce);
+  let streamController: ReadableStreamController<Uint8Array> | undefined;
+  const stream = new ReadableStream<Uint8Array>({
     start(controller) {
       if (typeof window === 'undefined') {
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/getReactServerComponent.client.ts:117]: Addressed in fb8f4905. Added an explicit invariant comment and close-on-non-browser path (`controller.close()`) to avoid any potential hanging stream if this function is called outside expected browser flow.

File: packages/react-on-rails-pro/src/getReactServerComponent.client.ts [hunk 15]
```diff
@@ -91,20 +98,33 @@ const fetchRSC = ({
...
     start(controller) {
+      // Browser-only by design (callers read from window.REACT_ON_RAILS_RSC_PAYLOADS).
+      // If called outside the browser, close immediately to avoid hanging streams.
       if (typeof window === 'undefined') {
+        controller.close();
         return;
       }
-      const handleChunk = (chunk: string) => {
-        controller.enqueue(chunk);
+      const handleChunk = (chunk: RSCPayloadChunk) => {
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/getReactServerComponent.client.ts:122]: Fixed in commit e01134d1 by adding a closed-state guard for preloaded payload handling. Late payload pushes after stream closure now no-op instead of enqueueing into a closed controller.

File: packages/react-on-rails-pro/src/injectRSCPayload.ts [hunk 16]
```diff
@@ -251,12 +254,38 @@ export default function injectRSCPayload(
...
+
+              // Contract: upstream stream emits NDJSON (one payload object per line).
+              for (const line of lines) {
+                const normalizedLine = line.trim();
+                if (normalizedLine !== '') {
+                  const payloadScript = createRSCPayloadChunk(normalizedLine, rscPayloadKey, sanitizedNonce);
+                  rscPayloadBuffers.push(Buffer.from(payloadScript));
+                }
+              }
+              scheduleFlush();
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/injectRSCPayload.ts]: Applied in commit e01134d1. scheduleFlush now runs only when at least one non-empty NDJSON line from the current chunk produced buffered payload output.

File: packages/react-on-rails-pro/src/transformRSCStreamAndReplayConsoleLogs.ts [hunk 17]
```diff
@@ -45,21 +45,9 @@ export default function transformRSCStreamAndReplayConsoleLogs(
...
-          .replace(/<\/script>$/i, '');
-        if (replayConsoleCode?.trim() !== '') {
-          const scriptElement = document.createElement('script');
-          if (sanitizedNonce) {
-            scriptElement.nonce = sanitizedNonce;
-          }
-          scriptElement.textContent = replayConsoleCode;
-          document.body.appendChild(scriptElement);
-        }
+        replayConsoleLog(consoleReplayScript, sanitizedNonce);
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/transformRSCStreamAndReplayConsoleLogs.ts:50]: Also fixed in commit e01134d1: transformRSCStreamAndReplayConsoleLogs now decodes Uint8Array chunks with stream mode and flushes remaining decoder state at end, so split multi-byte UTF-8 sequences are handled correctly on the fetch path.

File: packages/react-on-rails-pro/src/injectRSCPayload.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/injectRSCPayload.ts:59]: Addressed in d901c25f. createRSCPayloadChunk now validates each NDJSON line with JSON.parse before embedding it as a JS expression, and I added a regression test that verifies malformed input emits an error instead of injecting invalid script content.

File: packages/react-on-rails-pro/src/injectRSCPayload.ts [hunk 18]
```diff
@@ -251,12 +254,42 @@ export default function injectRSCPayload(
...
+                lastIncompleteLine +
+                (typeof chunk === 'string' ? chunk : decoder.decode(chunk, { stream: true }));
+              const lines = decodedChunk.split('\n');
+
+              if (!decodedChunk.endsWith('\n')) {
+                lastIncompleteLine = lines.pop() ?? '';
+              } else {
+                lastIncompleteLine = '';
+              }
+
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/injectRSCPayload.ts]: Addressed in d901c25f. I aligned this chunk-splitting logic with the transformRSCStream helper by always popping the final partial line.

File: packages/react-on-rails-pro/src/getReactServerComponent.client.ts [hunk 19]
```diff
@@ -114,33 +137,30 @@ const createRSCStreamFromArray = (payloads: string[]) => {
 
   if (typeof document !== 'undefined' && document.readyState === 'loading') {
     document.addEventListener('DOMContentLoaded', () => {
+      closed = true;
       streamController?.close();
     });
   } else {
+    closed = true;
     streamController?.close();
   }
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/getReactServerComponent.client.ts:148]: Addressed in d901c25f. I added the invariant comment on the already-loaded document path so the immediate close behavior is documented.

File: packages/react-on-rails-pro/src/injectRSCPayload.ts [hunk 20]
```diff
@@ -251,12 +259,37 @@ export default function injectRSCPayload(
...
+                  bufferedPayloadInThisChunk = true;
+                }
+              }
+              if (bufferedPayloadInThisChunk) {
+                scheduleFlush();
+              }
+            }
+            const finalChunk = lastIncompleteLine + decoder.decode();
+            if (finalChunk.trim() !== '') {
+              const payloadScript = createRSCPayloadChunk(finalChunk, rscPayloadKey, sanitizedNonce);
```
[the remark this answers is not in the document]
Comment (justin808) [reply packages/react-on-rails-pro/src/injectRSCPayload.ts:292]: Updated. The final buffered payload chunk now goes through the same trim-and-embed path as the in-loop lines, and I added a regression test that covers a CRLF-terminated final payload chunk.

**Comment by justin808** [comment]
Thanks for the detailed review and prioritization.

1) On fragile microtask tick counting: agreed in principle. In the current branch there are no remaining manual `await Promise.resolve()` ticks in this test path, so assertions are no longer coupled to promise-chain depth.

2) On post-unmount `finally` mutation / `cancelled`: this is intentional. `cancelled` only gates post-unmount logging while each mount gets a fresh router instance, and we keep a best-effort `cancelLoad()` call when available. I added/kept the inline intent comment in `clientHydrate.ts` to make this explicit.

3) On semver/type break callout: we are not narrowing `ssr` to `{ manifest?: unknown }` in the shipped type. It remains `ssr?: boolean | { manifest: unknown }` (`tanstack-router/types.ts`), so there is no new semver type break to call out in CHANGELOG for this PR.

Appreciate the careful read.
